docs/proposals/20-ha-control-planes.md

```diff
@@ -0,0 +1,1260 @@
+* During the switchover from `primary` to `hot-standby` if the `hot-standy` etcd is in process of applying delta snaphots or restoring from a new full snapshot then `hot-standby` should ensure that the backup-restore container sets the readiness probe to indicate that it is not ready yet causing additional downtime.
+</details>
+
+<details>
+<summary>Option-2: Perpetual Learner</summary>
+<img src="assets/activepassive-option2.png">
+
+In this option etcd cluster and learner facilities are leveraged. `etcd-druid` will bootstrap a cluster with one member. Once this member is ready to serve client requests then an additional `learner` will be added which joins the etcd cluster as a non-voting member. Learner will reject client reads and writes requests and the clients will have to reattempt. Typically switching to another member in a cluster post retries is provided out-of-the-box by etcd `clientv3`. The only `member` who is also the leader will serve all client requests.
+
+**Pros**
+* All the pros in `Option-1` are also applicable for this option.
+* Leaner will be continuously updated by the leader and will remain in-sync with the leader. Using a learner retains historical data and its ordering and is therefore better in that aspect as compared to `Option-2`.
+
+**Cons**
+* All the cons in `Option-1` are also applicable for this option.
+* `etcd-druid` will now have to additionally play an active role in managing members of an etcd cluster by add new members as `learner` and promoting `learner` to an active member if the leader is no longer available. This will increase the complexity in `etcd-druid`.
+* To prevent clients from re-attempting to reach the `learner`, `etcd-druid` will have to ensure that the label on the learner are set differently than the `leader`. This needs to be done every time there is a switch in the leader/learner.
+* Since etcd only allows addition of 1 learner at a time, this means that the HA setup can only have one failover etcd node, limiting its capability to have more than one `hot-standby`.
+</details>
+
+### Topology Spread Constraints evaluation and findings
+
+<details>
+<summary>Finding #1</summary>
+
+Consider the following setup:
+
+_Single zone & multiple nodes_
+
+<img src="assets/singlezone-multinode.png" style="width:200px;height:250px">
+
+When the constraints defined above are applied then the following was the findings:
+
+* With 3 replicas of etcd, all three got scheduled (one per node). This was a bit unexpected. As per the documentation if there are multiple constraints then they will be evaluated in conjunction. The first constraint should only allow 1 etcd pod per zone and the remaining 2 should not have been scheduled and should continue to be stuck in `pending` state. However all 3 etcd pods got scheduled and started successfully.
```

Comment:
This is an interesting case. Actuall, this is a correct behavior, according defintion of `global minimum` in the [official doc](https://kubernetes.io/docs/concepts/scheduling-eviction/topology-spread-constraints/). If you want only 1 node deployed in one zone, you may leverage field `minDomains`.